### PR TITLE
Task01 Артем Фадеев SPbU

### PR DESCRIPTION
**Pull request by artfadeev**
<details><summary>Локальный вывод</summary><p>

<pre>
Selected device: AMD Ryzen 5 5500U with Radeon Graphics         
Data generated for n=100000000!
Log:
Compilation started
Compilation done
Linking started
Linking done
Device build started
Device build done
Kernel <aplusb> was successfully vectorized (8)
Done.
Kernel average time: 0.0736003+-0.000166224 s
GFlops: 1.35869
VRAM bandwith: 7.59228 GB/s
Result data transfer time: 6.33333e-06+-1.31233e-06 s
VRAM -> RAM bandwidth: 14705.1 GB/s
</pre>

</p></details>

<details><summary>Вывод Github CI</summary><p>
<pre>
Selected device: Intel(R) Xeon(R) Platinum 8272CL CPU @ 2.60GHz
Data generated for n=100000000!
Log:
Compilation started
Compilation done
Linking started
Linking done
Device build started
Device build done
Kernel <aplusb> was successfully vectorized (16)
Done.
Kernel average time: 0.0525565+-0.000353786 s
GFlops: 1.90271
VRAM bandwith: 10.6322 GB/s
Result data transfer time: 4.41667e-06+-6.40095e-07 s
VRAM -> RAM bandwidth: 21086.5 GB/s
</pre>

</p></details>

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -141,34 +224,38 @@ int main() {
...
+        std::cout << "GFlops: " << ((double) n) / 1000000000 / t.lapAvg() << std::endl;
 
         // TODO 14 Рассчитайте используемую пропускную способность обращений к видеопамяти (в гигабайтах в секунду)
         // - Всего элементов в массивах по n штук
         // - Размер каждого элемента sizeof(float)=4 байта
         // - Обращений к видеопамяти 2*n*sizeof(float) байт на чтение и 1*n*sizeof(float) байт на запись, т.е. итого 3*n*sizeof(float) байт
         // - В гигабайте 1024*1024*1024 байт
         // - Среднее время выполнения кернела равно t.lapAvg() секунд
-        std::cout << "VRAM bandwidth: " << 0 << " GB/s" << std::endl;
+        std::cout << "VRAM bandwith: " << ((double) n*3*sizeof(float)) / (2L<<30) / t.lapAvg() << " GB/s"<<std::endl;
```
Comment (simiyutin) [inline src/main.cpp:235]: (2L<<30) это 2^31, то есть 2 гигабайта

File: src/main.cpp [hunk 2]
```diff
@@ -141,34 +224,38 @@ int main() {
...
     {
         timer t;
         for (unsigned int i = 0; i < 20; ++i) {
-            // clEnqueueReadBuffer...
+            clEnqueueReadBuffer(command_queue, c_buf, CL_TRUE, 0, n*sizeof(float), cs.data(), 0, nullptr, nullptr);
             t.nextLap();
         }
         std::cout << "Result data transfer time: " << t.lapAvg() << "+-" << t.lapStd() << " s" << std::endl;
-        std::cout << "VRAM -> RAM bandwidth: " << 0 << " GB/s" << std::endl;
+        std::cout << "VRAM -> RAM bandwidth: " << ((double) n*sizeof(float))/(2<<10)/(2<<20)/t.lapAvg() << " GB/s" << std::endl;
```
Comment (simiyutin) [inline src/main.cpp:247]: и здесь то же самое со сдвигом

File: src/main.cpp [hunk 3]
```diff
@@ -141,34 +224,38 @@ int main() {
...
-    //        }
-    //    }
+    for (unsigned int i = 0; i < n; ++i) {
+        if (cs[i] != as[i] + bs[i]) {
+            throw std::runtime_error("CPU and GPU results differ!");
+        }
+    }
 
+    // Cleanup
+    clReleaseCommandQueue(command_queue);
```
Comment (simiyutin) [inline src/main.cpp:258]: освободили не все служебные опенсл объекты и не освободили буферы

**Comment by artfadeev** [comment]
Спасибо, поправил.
 
<details><summary>Локальный вывод</summary><p>

<pre>
Selected device: AMD Ryzen 5 5500U with Radeon Graphics         
Data generated for n=100000000!
Log:
Compilation started
Compilation done
Linking started
Linking done
Device build started
Device build done
Kernel <aplusb> was successfully vectorized (8)
Done.
Kernel average time: 0.0775549+-0.00100624 s
GFlops: 1.28941
VRAM bandwith: 14.4103 GB/s
Result data transfer time: 8.33333e-06+-1.88562e-06 s
VRAM -> RAM bandwidth: 44703.5 GB/s
</pre>

</p></details>

<details><summary>Вывод Github CI</summary><p>
<pre>
Selected device: Intel(R) Xeon(R) Platinum 8370C CPU @ 2.80GHz
Data generated for n=100000000!
Log:
Compilation started
Compilation done
Linking started
Linking done
Device build started
Device build done
Kernel <aplusb> was successfully vectorized (16)
Done.
Kernel average time: 0.0434969+-0.000138974 s
GFlops: 2.29901
VRAM bandwith: 25.6935 GB/s
Result data transfer time: 3.41667e-06+-4.93007e-07 s
VRAM -> RAM bandwidth: 109033 GB/s
</pre>

</p></details>

**Comment by simiyutin** [comment]
Все хорошо, задача зачтена, 5/5 баллов :+1:

**Pull request closed without merging**